// @tresjs/eslint-config/dist/index.js
import antfu from "@antfu/eslint-config";

//#region src/rules/antfu.ts
const antfuConfig = {
	stylistic: true,
	ignores: [
		"**/dist/**",
		"sponsorkit/**",
		"node_modules/**",
		"coverage/**",
		"**/public/**"
	],
	formatters: {
		css: true,
		html: true
	}
};
var antfu_default = antfuConfig;

//#endregion
//#region src/rules/base.ts
const baseConfig = [{
	name: "tres:general",
	rules: {
		"curly": ["error", "all"],
		"style/function-call-spacing": ["error", "never"],
		"node/prefer-global/process": "off",
		"antfu/top-level-function": "off",
		"perfectionist/sort-imports": "off",
		"style/max-statements-per-line": "off",
		"jsdoc/check-alignment": "off"
	}
}, {
	name: "tres:vue",
	files: ["**/*.vue"],
	rules: {
		"vue/max-attributes-per-line": ["error", {
			singleline: { max: 10 },
			multiline: { max: 1 }
		}],
		"vue/singleline-html-element-content-newline": "off",
		"vue/html-self-closing": ["warn", { html: {
			void: "always",
			normal: "never"
		} }],
		"vue/attribute-hyphenation": "off",
		"vue/object-property-newline": ["error", { allowAllPropertiesOnSameLine: false }]
	}
}];
var base_default = baseConfig;

//#endregion
//#region src/rules/nuxt.ts
const nuxtConfig = [
	{
		name: "tres:nuxt:auto-imports",
		files: ["**/*.{js,ts,jsx,tsx,vue}"],
		rules: { "no-undef": "off" }
	},
	{
		name: "tres:nuxt:components",
		files: ["**/components/**/*.{js,ts,jsx,tsx,vue}"],
		rules: { "vue/multi-word-component-names": "warn" }
	},
	{
		name: "tres:nuxt:pages-layouts",
		files: ["**/{pages,layouts}/**/*.{js,ts,jsx,tsx,vue}"],
		rules: { "vue/no-multiple-template-root": "error" }
	}
];
var nuxt_default = nuxtConfig;

//#endregion
//#region src/index.ts
const tresLintConfig = (options = {}, ...configs) => {
	return antfu({
		...antfu_default,
		...options
	}, base_default, ...configs);
};

//#endregion
export { nuxt_default as nuxt, tresLintConfig };
