// @tresjs/eslint-config/dist/index.cjs
//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let __antfu_eslint_config = require("@antfu/eslint-config");
__antfu_eslint_config = __toESM(__antfu_eslint_config);

//#region src/rules/antfu.ts
const antfuConfig = {
	stylistic: true,
	ignores: [
		"**/dist/**",
		"sponsorkit/**",
		"node_modules/**",
		"coverage/**",
		"**/public/**"
	],
	formatters: {
		css: true,
		html: true
	}
};
var antfu_default = antfuConfig;

//#endregion
//#region src/rules/base.ts
const baseConfig = [{
	name: "tres:general",
	rules: {
		"curly": ["error", "all"],
		"style/function-call-spacing": ["error", "never"],
		"node/prefer-global/process": "off",
		"antfu/top-level-function": "off",
		"perfectionist/sort-imports": "off",
		"style/max-statements-per-line": "off",
		"jsdoc/check-alignment": "off"
	}
}, {
	name: "tres:vue",
	files: ["**/*.vue"],
	rules: {
		"vue/max-attributes-per-line": ["error", {
			singleline: { max: 10 },
			multiline: { max: 1 }
		}],
		"vue/singleline-html-element-content-newline": "off",
		"vue/html-self-closing": ["warn", { html: {
			void: "always",
			normal: "never"
		} }],
		"vue/attribute-hyphenation": "off",
		"vue/object-property-newline": ["error", { allowAllPropertiesOnSameLine: false }]
	}
}];
var base_default = baseConfig;

//#endregion
//#region src/rules/nuxt.ts
const nuxtConfig = [
	{
		name: "tres:nuxt:auto-imports",
		files: ["**/*.{js,ts,jsx,tsx,vue}"],
		rules: { "no-undef": "off" }
	},
	{
		name: "tres:nuxt:components",
		files: ["**/components/**/*.{js,ts,jsx,tsx,vue}"],
		rules: { "vue/multi-word-component-names": "warn" }
	},
	{
		name: "tres:nuxt:pages-layouts",
		files: ["**/{pages,layouts}/**/*.{js,ts,jsx,tsx,vue}"],
		rules: { "vue/no-multiple-template-root": "error" }
	}
];
var nuxt_default = nuxtConfig;

//#endregion
//#region src/index.ts
const tresLintConfig = (options = {}, ...configs) => {
	return (0, __antfu_eslint_config.default)({
		...antfu_default,
		...options
	}, base_default, ...configs);
};

//#endregion
exports.nuxt = nuxt_default;
exports.tresLintConfig = tresLintConfig;
